source/grammar/qasm3Parser.g4

```diff
@@ -0,0 +1,361 @@
+    | expression DOUBLE_PIPE logicalAndExpression
+    ;
+
+/**  Expression hierarchy for non-terminators. Adapted from ANTLR4 C
+  *  grammar: https://github.com/antlr/grammars-v4/blob/master/c/C.g4
+  * Order (first to last evaluation):
+    Terminator (including Parens),
+    Unary Op,
+    Multiplicative
+    Additive
+    Bit Shift
+    Comparison
+    Equality
+    Bit And
+    Exlusive Or (xOr)
+    Bit Or
+    Logical And
+    Logical Or
+**/
+
+logicalAndExpression
+    : bitOrExpression
+    | logicalAndExpression DOUBLE_AMPERSAND bitOrExpression
+    ;
+
+bitOrExpression
+    : xOrExpression
+    | bitOrExpression PIPE xOrExpression
+    ;
+
+xOrExpression
+    : bitAndExpression
+    | xOrExpression CARET bitAndExpression
+    ;
+
+bitAndExpression
+    : equalityExpression
+    | bitAndExpression AMPERSAND equalityExpression
+    ;
+
+equalityExpression
+    : comparisonExpression
+    | equalityExpression EqualityOperator comparisonExpression
+    ;
+
+comparisonExpression
+    : bitShiftExpression
+    | comparisonExpression ComparisonOperator bitShiftExpression
+    ;
+
+bitShiftExpression
+    : additiveExpression
+    | bitShiftExpression BitshiftOperator additiveExpression
+    ;
+
+additiveExpression
+    : multiplicativeExpression
+    | additiveExpression (PLUS | MINUS) multiplicativeExpression
+    ;
+
+multiplicativeExpression
+    // base case either terminator or unary
+    : powerExpression
+    | unaryExpression
+    | multiplicativeExpression (ASTERISK | SLASH | PERCENT) ( powerExpression | unaryExpression )
+    ;
+
+unaryExpression
+    : unaryOperator powerExpression
+    ;
+
+powerExpression
+    : expressionTerminator
+    | expressionTerminator DOUBLE_ASTERISK powerExpression
+    ;
+
+expressionTerminator
+    : Constant
+    | Integer
+    | RealNumber
+    | ImagNumber
+    | BooleanLiteral
+    | Identifier
+    | StringLiteral
+    | builtInCall
+    | externOrSubroutineCall
+    | timingIdentifier
+    | LPAREN expression RPAREN
+    | expressionTerminator LBRACKET expression RBRACKET
+    ;
+/** End expression hierarchy **/
+
+builtInCall: (BuiltinMath | castOperator) LPAREN expressionList RPAREN;
+
+castOperator: classicalType;
+
+expressionList: expression (COMMA expression)*;
+
+equalsExpression: EQUALS expression;
+assignmentOperator : EQUALS | CompoundAssignmentOperator;
+
+setDeclaration
+    : LBRACE expressionList RBRACE
+    | rangeDefinition
+    | Identifier
+    ;
+
+programBlock
+    : statement | controlDirective
+    | LBRACE ( statement | controlDirective )* RBRACE
+    ;
+
+branchingStatement: IF LPAREN expression RPAREN programBlock (ELSE programBlock)?;
+
+loopSignature
+    : FOR Identifier IN setDeclaration
+    | WHILE LPAREN expression RPAREN
+    ;
+
+loopStatement: loopSignature programBlock;
+endStatement: END SEMICOLON;
+returnStatement: RETURN (expression | quantumMeasurement)? SEMICOLON;
+
+controlDirective
+    : (BREAK| CONTINUE) SEMICOLON
+    | endStatement
```

Comment:
We have `endStatment` both here and in `statement` on line 30. It might confuse the parser, I am wondering if we should just put it in one place.

Reply:
Yeah, I guess it was probably an oversight in the original grammar that caused this.  It won't confuse ANTLR - it has a defined priority for which rule it tries to construct, so it'll reliably always pick the same production (though I don't remember exactly which it'll be).

Similar to other bits that actually change the grammar, we can fix it up in a follow-on.